### PR TITLE
Enable ETH fees

### LINKED ISSUE #1076: ETH fee in withdraw function
## Context

As discussed [here](https://github.com/sablier-labs/company-discussions/discussions/72#discussioncomment-11115807), we want to be able to monetize the Lockup streams by charging an ETH fee, since this is much easier than charging in USDC or the stream's token.

## Spec

- [x] Make all functions payable so that the UI can set a non-zero `msg.value`.
- [x] Allow anyone to call `withdrawFees` function, and always transfer the ETH fees to the contract admin.

Note: compared to the ETH fee in MerkleLockup, the fee in this proposal is *optional*. The goal is to keep the protocol free but monetize at the UI level — the Sablier UI would introduce a small ETH fee in the transactions signed by users.

### PR DESCRIPTION
**Pull request by andreivladbrg**
Depends on #1084 

Closes #1076  and https://github.com/sablier-labs/v2-core/issues/1090

### CODE REVIEW

**Comment by andreivladbrg** [comment]
> Shouldn't this PR be independent of https://github.com/sablier-labs/v2-core/pull/1084? Both are separate issues. One is related to batch functions which is primarily create related whereas this one is about withdraw function.

not necessarily. it’s not about one issue being directly related to the other, but rather about anticipating potential conflicts to avoid time wasted resolving them. you can see the independent changes related to eth fees, as the base branch is refactor/include-batch-core. after merging that branch, the base branch for this pr will automatically switch to staging.

we should merge PRs in chronological order

> It's a good practice to create PRs independent of each other. 

why? since all will eventually be merged into `staging`, it’s more about focusing on the changes specific to this issue, which is exactly what you see with the current base branch.

> If one takes time or requires more changes after review, the other PR can be reviewed and merged independently.

since it's just the two of us working on these changes, and this isn’t an open-source project with multiple contributors (where your approach would be preferred), i believe this approach is more efficient.
also, based on my experience with the unlock amount feature and the conflicts that arose, i can confidently say that creating PRs dependent on others is often more efficient.

#### scenario 1 (sequentially dependent PRs)

1. A <-- B <-- C <-- D
2.           B <-- B review
3.           C <-- rebase B review
4.           C <-- C review
5.           D <-- D review
          
#### scenario 2 (independent PRs with higher conflicts)

1. A <-- B 
2. A <-- C # changes similar files that B changes (conflicts warning)
3. A <-- D # changes similar files that B and C changes (conflicts warning)
4. B <-- B review
5. C <-- rebase B
6. D <-- rebase B
7. C <-- C review
8. D <-- rebase C
9. D <-- review D


as you can see, scenario 1 is better

**Comment by smol-ninja** [comment]
You made some good points. I gave it more thought, and I think what you said is accurate when the time complexity of reviewing, resolving comments, and merging all PRs is the same. In many cases, the time complexities are different. Making all PRs dependent on each other would mean it disallows us from reviewing, resolving, and merging them asynchronously. Even though some PRs may be reviewed and merged faster, they will continue to be blocked by the base PR, which may take more time than anticipated. Also, in the case of a chain of PRs, when changes are made to the base PR, that means making changes to all subsequent PRs.

Your point about rebasing is valid, but in my experience, we can identify such PRs/issues that are truly dependent on each other. In those cases, I agree with you. But generalizing it can slow us down in many cases. 

As an example, I rebased https://github.com/sablier-labs/v2-core/pull/1091 from staging in less than 30 minutes, and it can now also be reviewed and merged independently, which otherwise could stay open for as long as the other PRs are open. Another example is that https://github.com/sablier-labs/v2-core/pull/1087 can be worked upon, reviewed and merged independently of other PRs opened before it.

From my experience, I prefer creating independent PRs as much as possible, which you can review and merge independent of other PRs. This, imo, speeds up closing the issues faster without getting blocked by the base PR.

**Comment by andreivladbrg** [comment]
@smol-ninja i have marked the PR as ready to review, tho the fork tests are failing, we can ignore them for now, since we will remove the "periphery" contract in the upcoming PR

**Comment by smol-ninja** [comment]
Have you also included #1090 in this PR as discussed [here](https://github.com/sablier-labs/v2-core/issues/1090#issuecomment-2483369990)? If yes, can you please mention it in the description?

**Comment by andreivladbrg** [comment]
> Have you also included #1090 in this PR as discussed [here](https://github.com/sablier-labs/v2-core/issues/1090#issuecomment-2483369990)? If yes, can you please mention it in the description?

not yet, facing some issue with it, we might be able to implement it, due to requiring all functions within `calls` array to be payable as well

**Comment by andreivladbrg** [comment]
@smol-ninja Thank you for your detailed reply. 

> You made some good points. I gave it more thought, and I think what you said is accurate when the time complexity of reviewing, resolving comments, and merging all PRs is the same.

Indeed, complexity plays a significant role in this discussion.

> In many cases, the time complexities are different

You are correct that PR complexities can vary. IMO, the sequential approach works best for handling highly complex PRs, whereas the parallel approach is more suited for small changes.

> Making all PRs dependent on each other would mean it disallows us from reviewing

That's incorrect and I disagree with your assertion regarding the base branch of the PR. As I mentioned in my initial comment, what truly matters is the base branch the PR is raised against. For example, if I start from `refactor/include-batch-core` and create a PR with the base branch set as `refactor/include-batch-core`, you will only review the changes specific to `feat/eth-fee`. This is because the git diff will display only those changes.


> Even though some PRs may be reviewed and merged faster, they will continue to be blocked by the base PR

I don’t quite understand why this point is relevant. Whether a PR is merged into staging after 1 day or after 5 days doesn’t seem to matter as long as all PRs ultimately need to be merged into `staging` for the code to be audited.

Let me give an example: Suppose everything must be merged into staging. I create branches A and B, both involving complex changes and targeting the same files and lines (guaranteeing conflicts). Branch B is created after Branch A, but the PR for Branch B is raised first. The review processes for both start simultaneously. If PR A takes 1 day for review and changes, and PR B takes 2 days, merging PR A immediately upon completion would require PR B to be rebased and conflicts resolved, adding another day (3 days total).

In contrast, if Branch B were created sequentially from Branch A, rebasing the review changes of PR A into Branch B would only take half a day. In this case, having both branches on staging would take 2.5 days instead of 3 days. For more complex PRs, I still believe the sequential method is more efficient.

Another argument is that, for us (contract development), we don’t need specific features merged into the `main` branch as soon as possible, unlike other repositories (like `v2-interfaces`). In that case, I completely agree with you that parallel PRs are the way to go. However, for us, since we release our code after it has been audited, we don’t need to merge into the `main` or `staging` branch immediately. I hope I have made myself clear

> which may take more time than anticipated. Also, in the case of a chain of PRs, when changes are made to the base PR, that means making changes to all subsequent PRs.

As mentioned above, in the end, it might take more time to have all changes (what really matters) merged into staging 

> As an example, I rebased https://github.com/sablier-labs/v2-core/pull/1091 from staging in less than 30 minutes, and it can now also be reviewed and merged independently, 


This is a small PR, and it’s understandable that it didn’t take much time (which I agree is not that relevant to handle in sequence). However, I come from the singleton and unlock amounts PR, where rebasing took a lot of time (1 day).

> which otherwise could stay open for as long as the other PRs are open. Another example is that https://github.com/sablier-labs/v2-core/pull/1087 can be worked upon, reviewed and merged independently of other PRs opened before it.

IMO, as said above, that's not that relevant when we merge it if we don't have all changes ready for auditing. But I understand the argument since it is a small PR

> From my experience, its easier to deal with independent PRs which I also prefer as much as possible, which you can review and merge independent of other PRs. 

And from my experience, when it comes to more complex PRs, it’s generally easier to handle them sequentially.”

> speeds up closing the issues faster without getting blocked by the base PR

not necessarily

---

PS: You don’t need to reply to this now. We already discussed on Slack using the parallel option

**Comment by andreivladbrg** [comment]
hmm, a weird error we are facing when compiling

image

not sure yet what causes it

**Comment by smol-ninja** [comment]
> not sure yet what causes it

I will investigate. Feel free to go to bed :))

File: src/core/interfaces/ISablierLockupBase.sol
Comment (smol-ninja) [inline src/core/interfaces/ISablierLockupBase.sol:325]: This can be removed since there is no concept of Sablier fee yet.

File: src/core/libraries/Errors.sol
Comment (smol-ninja) [inline src/core/libraries/Errors.sol:172]: Unsued.

File: test/Base.t.sol
Comment (smol-ninja) [inline test/Base.t.sol:96]: Lets revert it since this is going to be removed in https://github.com/sablier-labs/v2-core/pull/1089

File: test/Base.t.sol
Comment (smol-ninja) [inline test/Base.t.sol:69]: Why does `recipientGood` need ETH?

File: test/core/integration/Integration.t.sol
Comment (smol-ninja) [inline test/core/integration/Integration.t.sol:354]: Why not put this modifier in the `Modifiers.sol` contract and override it like other examples?

File: test/core/integration/Integration.t.sol
Comment (smol-ninja) [inline test/core/integration/Integration.t.sol:276]: IMO these functions are overkill. If the `value` is 0, then it is equivalent to to not passing any value at all.

```solidity
lockup.burn{ value: 0 }(streamId);
```

is same as 

```solidity
lockup.burn(streamId);
```

File: test/core/integration/Integration.t.sol
Comment (smol-ninja) [inline test/core/integration/Integration.t.sol:378]: I'd be in favour of removing these helpers and use the function itself where it is needed with a non-zero `value`

File: test/core/integration/concrete/lockup-base/batch/batch.t.sol
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/batch/batch.t.sol:38]: We can replace it with 

```solidity
lockup.batch(calls);
```

File: test/core/integration/concrete/lockup-base/burn/burn.t.sol
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/burn/burn.t.sol:22]: We should revert it back.

File: test/core/integration/concrete/lockup-base/withdraw-fees/withdrawFees.tree
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/withdraw-fees/withdrawFees.tree:15]: I feel this is redundant branch. We should be able to remove it.

File: test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol:66]: The advantage of using `lockup.withdraw` with explicit `value` is that we can differentiate which tests is with 0 value and which are not. In the current approach, its hard to differentiate between the two by looking at the code.

File: test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree:9]: this is not being tested in the test contract so no need to mention that it should charge the fee.

File: test/core/integration/concrete/lockup-base/withdraw/withdraw.t.sol
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/withdraw/withdraw.t.sol:173]: Not included in the tree.

File: test/core/integration/concrete/lockup-base/withdraw/withdraw.tree
Comment (smol-ninja) [inline test/core/integration/concrete/lockup-base/withdraw/withdraw.tree:65]: I could not find where its tested in the test code.

File: test/core/integration/fuzz/lockup-base/withdraw.t.sol
Comment (smol-ninja) [inline test/core/integration/fuzz/lockup-base/withdraw.t.sol:15]: Why does `caller` have to be `payable`?

File: test/core/invariant/handlers/LockupHandler.sol
Comment (smol-ninja) [inline test/core/invariant/handlers/LockupHandler.sol:170]: Why do we need a non-zero `SABLIER_FEE` in invariant if fee is not a part of any invariant statement?

**Review by smol-ninja — CHANGES_REQUESTED** [review]
Great work @andreivladbrg. Left some comments below. I have also included a commit similar to https://github.com/sablier-labs/airdrops/pull/5/commits/0611cf9bf877c7665b12f0c861fd87ce5d139afd which renamed sablier fee to simply fee.

File: src/core/interfaces/ISablierLockupBase.sol
Comment (smol-ninja) [reply src/core/interfaces/ISablierLockupBase.sol]: https://github.com/sablier-labs/v2-core/pull/1087/commits/1f8d30b76893c80e77a6c6a1ea24737fa32b81ba

File: src/core/libraries/Errors.sol
Comment (smol-ninja) [reply src/core/libraries/Errors.sol]: https://github.com/sablier-labs/v2-core/pull/1087/commits/1f8d30b76893c80e77a6c6a1ea24737fa32b81ba

File: test/Base.t.sol
Comment (andreivladbrg) [reply test/Base.t.sol:69]: Because it is used to call withdraw in some tests, and if it has a zero balance it would revert

File: test/core/integration/Integration.t.sol
Comment (andreivladbrg) [reply test/core/integration/Integration.t.sol:356]: My rationale is that the modifiers placed in that file, are the BTT modifiers

File: test/core/integration/Integration.t.sol
Comment (andreivladbrg) [reply test/core/integration/Integration.t.sol:380]: Fair enough, agree

File: test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol
Comment (andreivladbrg) [reply test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol:66]: All calls to withdraw have value greater than zero. So not sure what you are referring to

File: test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree
Comment (andreivladbrg) [reply test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree:9]: That is incorrect, the function that runs the balanceTest modifier is being is used

File: test/core/invariant/handlers/LockupHandler.sol
Comment (andreivladbrg) [reply test/core/invariant/handlers/LockupHandler.sol:170]: Good point, we don't need it

File: test/core/integration/concrete/lockup-base/withdraw/withdraw.t.sol
Comment (andreivladbrg) [reply test/core/integration/concrete/lockup-base/withdraw/withdraw.t.sol]: We shall remove it

File: test/core/integration/concrete/lockup-base/withdraw/withdraw.tree
Comment (andreivladbrg) [reply test/core/integration/concrete/lockup-base/withdraw/withdraw.tree:65]: It is by using withdrawWithBalTest function

File: test/Base.t.sol
Comment (smol-ninja) [reply test/Base.t.sol:69]: If it calls to `withdraw` then thats a reentrancy test. Don't we use `reentrantRecipient` for that? I thought `recipientGood` is the good recipient that does not try to reenter.

File: test/core/integration/Integration.t.sol
Comment (smol-ninja) [reply test/core/integration/Integration.t.sol:356]: The original decision was to dump all modifiers to a single file. https://github.com/sablier-labs/v2-core/issues/1028#issuecomment-2315376028. And override whenever necessary. It was not related to BTT.

File: test/core/integration/Integration.t.sol
Comment (smol-ninja) [reply test/core/integration/Integration.t.sol:278]: > Do you find it useful to test the ETH balance increase in all functions

No I don't. One concrete test for `withdrawFee` is enough imo as its a fairly simple logic.

File: test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol
Comment (smol-ninja) [reply test/core/integration/concrete/lockup-base/withdraw-hooks/withdrawHooks.t.sol:66]: I see. I thought it is only in `withdraw` tests and not all calls. In that case, can we please have explicitness instead of using these helper functions? May be have non zero fee in withdraw test but 0 feel at all other places in Concrete.

File: test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree
Comment (smol-ninja) [reply test/core/integration/concrete/lockup-base/withdraw-multiple/withdrawMultiple.tree:9]: Oh. Usually we test assertions at the end of the logic. So thats where I looked for it.

File: test/core/integration/concrete/lockup-base/withdraw/withdraw.tree
Comment (smol-ninja) [reply test/core/integration/concrete/lockup-base/withdraw/withdraw.tree:65]: Same as https://github.com/sablier-labs/v2-core/pull/1087#discussion_r1851757246.

**Comment by PaulRBerg** [comment]
Closing in favor of https://github.com/sablier-labs/v2-core/pull/1093.

**Pull request closed without merging**
